File: B-Learning-Demo/backend/app/models/submission.py
"""
Assignment Submission Models (Pydantic Schemas)
================================================

Schemas cho Assignment Submissions (liên quan nhiều bảng).

Author: Nguyễn Văn Kiệt - CNTT1-K63
"""

from pydantic import BaseModel, Field
from typing import Optional
from datetime import datetime
from decimal import Decimal


class SubmissionListItem(BaseModel):
    """
    Schema cho item trong danh sách submissions.
    Kết hợp dữ liệu từ nhiều bảng:
    - AssignmentSubmission
    - User (student info)
    - Assignment
    - Course
    """
    # From AssignmentSubmission
    assignment_submission_id: str
    submission_number: int
    submitted_at: datetime
    content: Optional[str] = None
    is_late: bool
    days_late: int
    status: str
    auto_score: Optional[Decimal] = None
    manual_score: Optional[Decimal] = None
    final_score: Optional[Decimal] = None
    penalty_applied: Decimal
    feedback: Optional[str] = None
    graded_at: Optional[datetime] = None

    # From User (student)
    student_id: str
    student_email: str
    student_name: str  # first_name + last_name

    # From Assignment
    assignment_id: str
    assignment_title: str
    assignment_type: str
    max_points: Decimal
    due_date: datetime

    # From Course
    course_id: str
    course_code: str
    course_title: str

    # From User (grader) - optional
    graded_by_name: Optional[str] = None

    class Config:
        from_attributes = True


class SubmissionDetail(SubmissionListItem):
    """
    Schema chi tiết submission (bao gồm thêm thông tin)
    """
    # Additional fields from AssignmentSubmission
    file_urls: Optional[dict] = None
    code_submission: Optional[str] = None
    rubric_scores: Optional[dict] = None

    # Additional fields from Assignment
    assignment_description: str
    assignment_instructions: Optional[str] = None
    late_submission_allowed: bool
    late_penalty_percent: Decimal

    # Additional fields from Course
    course_description: Optional[str] = None


class SubmissionListResponse(BaseModel):
    """Response cho danh sách submissions"""
    total: int
    submissions: list[SubmissionListItem]


class GradeSubmissionRequest(BaseModel):
    """Request body cho việc chấm điểm"""
    manual_score: Decimal = Field(..., ge=0, description="Điểm manual (>= 0)")
    feedback: Optional[str] = Field(None, description="Nhận xét của giảng viên")
    rubric_scores: Optional[dict] = Field(None, description="Điểm theo rubric (JSON)")


class SubmissionStats(BaseModel):
    """Thống kê submission"""
    total_submissions: int
    submitted: int
    graded: int
    pending: int
    average_score: Optional[Decimal] = None
    late_submissions: int
    on_time_submissions: int

File: B-Learning-Demo/frontend/types/submission.ts
/**
 * Submission Types
 * =================
 *
 * TypeScript types cho Assignment Submissions.
 */

export interface SubmissionListItem {
  // From AssignmentSubmission
  assignment_submission_id: string;
  submission_number: number;
  submitted_at: string;
  content?: string | null;
  is_late: boolean;
  days_late: number;
  status: 'DRAFT' | 'SUBMITTED' | 'GRADING' | 'GRADED' | 'RETURNED';
  auto_score?: number | null;
  manual_score?: number | null;
  final_score?: number | null;
  penalty_applied: number;
  feedback?: string | null;
  graded_at?: string | null;

  // From User (student)
  student_id: string;
  student_email: string;
  student_name: string;

  // From Assignment
  assignment_id: string;
  assignment_title: string;
  assignment_type: 'ESSAY' | 'CODE' | 'FILE_UPLOAD' | 'PROBLEM_SET' | 'PROJECT';
  max_points: number;
  due_date: string;

  // From Course
  course_id: string;
  course_code: string;
  course_title: string;

  // From User (grader)
  graded_by_name?: string | null;
}

export interface SubmissionDetail extends SubmissionListItem {
  // Additional fields
  file_urls?: any | null;
  code_submission?: string | null;
  rubric_scores?: any | null;
  assignment_description: string;
  assignment_instructions?: string | null;
  late_submission_allowed: boolean;
  late_penalty_percent: number;
  course_description?: string | null;
}

export interface SubmissionListResponse {
  total: number;
  submissions: SubmissionListItem[];
}

export interface GradeSubmissionRequest {
  manual_score: number;
  feedback?: string;
  rubric_scores?: any;
}

export interface SubmissionStats {
  total_submissions: number;
  submitted: number;
  graded: number;
  pending: number;
  average_score?: number | null;
  late_submissions: number;
  on_time_submissions: number;
}
